iocore/utils/PropertyBlock.h

```diff
@@ -0,0 +1,254 @@
+  ///
+  /** If you have to override the new/delete of the derived class be sure to
+   * call these PropertyBlock::alloc(sz) and PropertyBlock::free(ptr)
+   */
+
+  /// allocate a new object with properties
+  void *
+  operator new(size_t size)
+  {
+    // allocate one block
+    const size_t alloc_size = size + s_properties_total_size + (s_bits_size + 7) / 8;
+    void *ptr               = ::operator new(alloc_size);
+    memset(ptr, 0, alloc_size);
+
+    return ptr;
+  }
+
+  PropertyBlock() { s_instance_count++; }
+  // we have to all propBlockInit after the Derived contructor. IDK how to do that implicitly.
+
+  ~PropertyBlock()
+  {
+    s_instance_count--;
+    propBlockDestroy();
+  }
+
+  static bool
+  resetSchema()
+  {
+    if (s_instance_count > 0) {
+      // free instances before calling this so we don't leak memory
+      return false;
+    }
+    s_properties_total_size = 0;
+    s_bits_size             = status_bits::num_status_bits;
+    getSchema().clear();
+    return true;
+  }
+
+protected:
+  void
+  propBlockInit()
+  {
+    // is already constructed? prevents recursion
+    if (propGetBit(initialized)) {
+      return;
+    }
+    propPutBit(initialized, true);
+    for (Block &block : getSchema()) {
+      if (block.m_init) {
+        block.m_init(static_cast<Derived_t *>(this), propBlockGetPtr(block.m_offset));
+      }
+    }
+  }
+
+  void
+  propBlockDestroy()
+  {
+    // is already destroyed? prevents recursion
+    if (propGetBit(destroyed)) {
+      return;
+    }
+    propPutBit(destroyed, true);
+    for (Block &block : getSchema()) {
+      if (block.m_destroy) {
+        block.m_destroy(static_cast<Derived_t *>(this), propBlockGetPtr(block.m_offset));
+      }
+    }
+  }
+
+  /// returns a pointer at the given offset
+  inline void *
+  propBlockGetPtr(Offset_t offset) const
+  {
+    return ((char *)this) + offset;
+  }
+
+  static bool
+  hasReadAccess(const PropertyBlock *pb)
+  {
+    // Derived_t should override this static member function if it uses locking
+    return true;
+  }
+  static bool
+  hasWriteAccess(const PropertyBlock *pb)
+  {
+    // Derived_t should override this static member function if it uses locking
+    return true;
+  }
+
+private:
+  /// total size of all declared property structures
+  static uint32_t s_properties_total_size;
+
+  enum status_bits {
+    initialized, // true if the propBlock is initialized, false if destroyed.
+    destroyed,
+    num_status_bits
+  };
+  /// total bits of all declared propertyBits
+  static uint32_t s_bits_size;
+
+  /// the number of class instances in use.
+  static std::atomic_uint s_instance_count;
+
+  struct Block {
+    Offset_t m_offset;
+    PropertyFunc_t *m_init;    ///< called after new
+    PropertyFunc_t *m_destroy; ///< called before delete
+
+    // if we need to save and load, then we should add names to the blocks.
+  };
+  static std::vector<Block> &
+  getSchema()
+  {
+    static std::vector<Block> s_schema;
```

Comment:
Any benefit to isolating this inside a method? Avoids have to explicitly declare the storage?

Reply:
I couldn't figure out how to define it with all the templates going on. A template of templates didn't work. And I can't define Block independently.